[Security] Rework the remember me system

wouterj opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fixes part of #39308
| License       | MIT
| Doc PR        | tbd

As I said in #39308, I want to change the remember me system in Symfony 5.3. The remember me system has a couple big "problems":

1. **It's hardwired into some Security classes** like `ContextListener`. The `RememberMeFactory` adds a `setRememberMe()` method call to the DI config and the context listener calls methods on this. This is very coupled, instead of the decoupled nature of the rest of security.
2. **Conditional conditions are combined with cookie creation in one class**. This is especially hard in e.g. 2FA (where setting the cookie should be done after 2FA is completed, which is currently near impossible as it's directly bound to the conditional of being called after logging in).

The changes
---

* The first commits harden the current functional test suite of remember me, to avoid breaking it.
* I discovered a lot of similarity between remember me tokens and login links. That's why I've extracted the shared logic into a generic `SignatureHasher` in the 3rd commit.
* I then remodelled `RememberMeAuthenticator` to the login link system, which I think improves a lot and at least improves problem (2) - as the conditionals (`RememberMeAuthenticator`) is split from the cookie creation (`RememberMeHandlerInterface`).
* Finally, I added a new event (`TokenDeauthenticatedEvent`) to the `ContextListener` to avoid direct coupling - solving problem (1).

This removes any usage of remember me services, which can be deprecated along with the rest of the security system.

Usage
---

As with the authenticator manager: **Nothing changes in the configuration**

Usage of persistent token providers has been improved. First, configuration is provided (setting up services is no longer needed):
```yaml
# before
services:
    Symfony\Bridge\Doctrine\Security\RememberMe\DoctrineTokenProvider:
        autowire: true

security:
    firewalls:
        main:
            remember_me:
                # ...
                token_provider: 'Symfony\Bridge\Doctrine\Security\RememberMe\DoctrineTokenProvider'

# after
security:
    firewalls:
        main:
            remember_me:
                # ...
                token_provider:
                    doctrine: true
```

Furthermore, a schema listener is created. Whenever the doctrine token provider is used, `make:migration`/`doctrine:schema:update` will automatically create the required table.

Some advanced usage of Remember me is also improved a lot (there is no real "before" here, consider looking at scheb/2fa to get an idea of the before). A few use-cases I took into account:

* If you ever need to **programmatically create a remember me cookie**, you can autowire `RememberMeHandlerInterface` and use `createRememberMeCookie($user)`. This will make sure the remember me cookie is set on the final response (using the `ResponseListener`)
* The `RememberMeListener` previously was responsible for both determining if a cookie must be set and setting the cookie. This is now split in 2 listeners (checking is done by `RememberMeConditionsListener`). If `RememberMeBadge` is enabled, the cookie is set and otherwise it isn't. This allows e.g. SchebTwoFactorBundle to create a listener that catches whether remember me was requested, but suppress it until the 2nd factor is completed.

Todo
---

* [x] Update UPGRADE and CHANGELOG
* [x] Show before/after examples
* [x] Investigate the conditional event registering of `ContextListener`. This forces to inject both the firewall and the global event dispatcher at the moment.
* Make sure old remember me tokens still function. As remember me tokens are long lived, we may need to provide backwards compatibility for at least Symfony 6.x. **Update: it was decided to not include this for now: https://github.com/symfony/symfony/pull/40145#issuecomment-785819607**

cc @scheb @weaverryan as you both initiated this PR by sharing the problems with the current design.

wouterj committed:
[Security] Rework the remember me system

stof commented:
> * This removes the capability of `PersistentTokenRememberMeService`. Do we want to reintroduce this, or do we think that this is secure enough as-is (considering the same ideas as with the login link).

`PersistentTokenRememberMeService` is the only way to be able to invalidate tokens. And unlike login links, the remember tokens cannot be short-lived as that would defeat their purpose. So I suspect that removing the feature might not be OK (even though it was never properly documented and so requires digging into the code to discover it, making it an advanced features).

scheb commented on src/Symfony/Component/Security/Http/Event/TokenDeauthenticatedEvent.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+final class TokenDeauthenticatedEvent extends Event
```
Not 100% sure about it, you might want to add this event to `RegisterGlobalSecurityEventListenersPass`.

scheb commented on src/Symfony/Component/Security/Http/EventListener/RememberMeListener.php:
```diff
@@ -53,27 +71,61 @@ public function onSuccessfulLogin(LoginSuccessEvent $event): void
             return;
         }
 
-        $this->rememberMeServices->loginSuccess($event->getRequest(), $event->getResponse(), $event->getAuthenticatedToken());
```
This call was previously checking if the "remember me" flag/checkbox was set and therefore decided if the cookie is added or not. I'm wondering, where this logic would live after the rework. Can't find anything in the current changeset (I guess because it's WIP).

I'm especially interested in that aspect, because 2FA needs a way to prevent the default behavior (remember me cookie set right after login).

wouterj replied:
Very good point, I completely forgot to implement all if statements in `AbstractRememberMeServices::loginSuccess()`

wouterj commented:
> PersistentTokenRememberMeService is the only way to be able to invalidate tokens. And unlike login links, the remember tokens cannot be short-lived as that would defeat their purpose. So I suspect that removing the feature might not be OK (even though it was never properly documented and so requires digging into the code to discover it, making it an advanced features).

You can invalidate tokens by including properties in the signature. If any signature property changes, all previously made remember me tokens are invalidated. The difference being short-lived login links vs long-lived remember me tokens is an interesting one! I see the current persistent tokens are using the logic as explained in https://stackoverflow.com/a/244907/1149495 I'll see if I can find other examples/references about persistent remember me cookies and whether using signature properties can fulfill this (or we need to reintroduce a persistent token handler).

weaverryan commented:
> `PersistentTokenRememberMeService` is the only way to be able to invalidate tokens

As Wouter said, the signature properties could be used to invalidate tokens... but it could only be used to invalidate *all* tokens (e.g. by adding some `lastLogoutAt` property to `User` and having that in the signature properties). If I logged in at work and at home (using remember me on both) and then logged out at work, there would not be a way to fully invalidate JUST the remember me cookie for my work computer. We would only be able to invalidate *all* current remember me tokens. This is actually a UX problem with removing `PersistentTokenRememberMeService` for companies that want a more secure setup.

However, I will also say that the new system will make it easier for "everyone else" to have a more secure remember me system: we can document how to setup a system that would truly invalidate remember me tokens (via the signature properties) on log out (though, again, it would invalidate all remember me tokens for that user).

weaverryan commented on src/Symfony/Component/Security/Http/Firewall/ContextListener.php:
```diff
@@ -86,6 +89,7 @@ public function supports(Request $request): ?bool
      */
     public function authenticate(RequestEvent $event)
     {
+        // @todo do we need conditionally register?
```
I think so? I guess this is registered this way for performance? So that we don't bother doing our cookie-setting logic if there was no authentication during this request?

This is the only place now where we need the global event dispatcher now. This could potentially be done in a different way. The rough idea would be something like this:

A) Have a listener on `kernel.response` that is *always* registered when security is active (or, more accurately, when there is at least one stateful firewall)

B) Inject a service locator of the `ContextListener` for all firewalls... and something to be able to determine the current firewall (`FirewallContext` or whatever allows us to do that).

C) Then, it would look to see if the currently matched firewall has a `ContextListener`. If it does, it would call `onKernelResponse()` on it (which maybe has a different name). Internally, in `ContextListener`, instead of conditionally registering a global event listener, it could just set a flag - e.g .`$this->didAuthenticate = true` here and then reset back to  `$this->didAuthenticate = false` later (where the unregistering currently happens).

Of course, this doesn't take into account BC, but it seems like the cleaner way to do things.

wouterj replied:
We already keep state whether or not this listener is registered: https://github.com/symfony/symfony/blob/f50e6afd7dd90e5a09e3e2ef81a2b10a8185e812/src/Symfony/Component/Security/Http/Firewall/ContextListener.php#L91

I am/was wondering whether we could use this in `onKernelResponse()` (we already do) and just return early instead of magically registering this on the fly.

wouterj replied:
I've changed this to instead use request attributes. The current change is fully backwards compatible and the "easiest" imho.

For reference, this code is working this way to fix standalone usage of Security (in e.g. Silex). see https://github.com/symfony/symfony/pull/8946 and https://github.com/symfony/symfony/issues/8244 for some background.

stof replied:
Doing dynamic registration of a listener in the main event dispatcher service as a performance optimization is likely to backfire: the event dispatcher service is optimize to compute all the sorting of listeners by priority at build time rather than during usage. But if you dynamically register a listener, it requires redoing all the sorting as it invalidates the current sorted state.

weaverryan commented on src/Symfony/Component/Security/Http/Firewall/ContextListener.php:
```diff
@@ -128,10 +132,17 @@ public function authenticate(RequestEvent $event)
+
+                if ($this->rememberMeServices) {
+                    $this->rememberMeServices->loginFail($request);
+                }
```
Is this now a deprecated path? Only the old system would use this?

wouterj replied:
Yes, `$this->rememberMeServices` always is `null` in the new system.

weaverryan commented on src/Symfony/Component/Security/Http/Event/TokenDeauthenticatedEvent.php:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
```
Probably still a TODO for you, but this needs a nice description of its purpose, especially as it's very similar in name to `DeauthenticatedEvent` - there may be some confusion.

wouterj replied:
Oh definitely, haven't looked at a single PHPdoc yet.

weaverryan reviewed (commented):
This looks incredibly promising!

wouterj commented:
PR and description updated:

* I've reintroduced a persistent remember me implementation. I also tried to improve its use (default config + schema listener for automatic migration creation)
* I've reimplemented the conditions to set a remember me cookie (request parameter) and added a special way to opt-out.
* I've added lots of tests, CHANGELOG and PHPdoc.

stof commented on src/Symfony/Bridge/Doctrine/Security/RememberMe/DoctrineTokenProvider.php:
```diff
@@ -40,11 +41,13 @@
+    private const TABLE = 'rememberme_token';
 
-    public function __construct(Connection $conn)
+    private $connection;
```
renaming the private property will increase the conflicts between branches. So it should either be done in 4.4 or not be done IMO.

wouterj replied:
Ok, reverted

stof commented on src/Symfony/Bridge/Doctrine/Security/RememberMe/DoctrineTokenProvider.php:
```diff
@@ -53,11 +56,10 @@ public function __construct(Connection $conn)
         // the alias for lastUsed works around case insensitivity in PostgreSQL
-        $sql = 'SELECT class, username, value, lastUsed AS last_used'
-            .' FROM rememberme_token WHERE series=:series';
+        $sql = 'SELECT class, username, value, lastUsed AS last_used FROM '.self::TABLE.' WHERE series=:series';
```
I'd rather keep the SQL query as a raw literal rather than concatenating a private constant in it (which won't make it configurable anyway). That makes inspecting SQL queries easier.

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -148,16 +178,38 @@ public function addConfiguration(NodeDefinition $node)
+                ->arrayNode('doctrine')
+                    ->canBeEnabled()
+                    ->children()
+                        ->scalarNode('connection')->defaultValue('default')->end()
```
instead of hardcoding `default` as connection name (which is not guaranteed to be the name of the default DBAL connection as that name is customizable), the default value should be `null`. And when set to `null`, you should use the `database_connection` id for the reference, which is the alias pointing to the default DBAL connection.

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -148,16 +178,38 @@ public function addConfiguration(NodeDefinition $node)
+                ->end()
+                ->children()
+                    ->scalarNode('service')->end();
+        if (class_exists(DoctrineTokenProvider::class)) {
```
defining config nodes dynamically is something that is being removed in #40140 so I think this should be avoided. The configuration tree should always have the nodes defined here

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -94,31 +98,57 @@ public function create(ContainerBuilder $container, string $id, array $config, ?
 
-        // create remember me services (which manage the remember me cookies)
-        $this->createRememberMeServices($container, $firewallName, $templateId, [new Reference($userProviderId)], $config);
+        $options = $config + $this->options;
```
this merging is not needed. The configuration processing already registers default values for all options.

wouterj commented:
PR updated, including:
* functional tests for the persistent token handler
* a split of `RememberMeListener` (@scheb came with the great idea of splitting checking conditions and creating cookies in 2 listeners, so that 2fa can intercept the conditions check and suppress creating the cookie)
* removing the need to dynamically register `ContextListener::onKernelResponse()` by using request attributes. I have no idea how to properly deprecate it however.

_(build failures appear unrelated ref #40304)_

I didn't do this last original todo:

> Make sure old remember me tokens still function. As remember me tokens are long lived, we must provide backwards compatibility for at least Symfony 6.x imho

I'm more and more convinced that it isn't worth the complexity and instead it's "ok" to reset remember me tokens after an application upgraded to the new system. If you disagree: please let me know :)

weaverryan commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/RememberMeBadge.php:
```diff
@@ -30,6 +25,36 @@
+     *
+     * In most cases, you should not use this in your authenticator, but
+     * instead rely on {@see RememberMeConditionsListener} to enable it
+     * for you.
```
While this is true, it IS perfectly valid to call this in your authenticator to explicitly force remember me. I don't want to overly discourage people from this :)

wouterj replied:
I did want to discourage people from doing this :) But after reading the referenced Oauth2 issue, I agree that we shouldn't overly discourage using this. I've updated the sentence

wouterj commented:
This PR is now also functionally tested by the [SchebTwoFactorBundle with great success](https://github.com/scheb/2fa/pull/61). I consider this PR finished and ready for review/merge. The failing tests appear unrelated to me.

chalasr commented on src/Symfony/Bridge/Doctrine/SchemaListener/RememberMeTokenProviderDoctrineSchemaSubscriber.php:
```diff
@@ -0,0 +1,64 @@
+    private $rememberMeHandlers;
+
+    /**
+     * @param iterable|RememberMeHandlerInterface[] $doctrineTokenProviders
```
```suggestion
     * @param iterable|RememberMeHandlerInterface[] $rememberMeHandlers
```

chalasr commented on src/Symfony/Bundle/SecurityBundle/Security/FirewallAwareTrait.php:
```diff
@@ -0,0 +1,52 @@
+        if (!$this->locator->has($firewallName)) {
+            $message = 'No '.$getServiceIdentifier().' found for this firewall.';
+            if (\defined(static::class.'::FIREWALL_OPTION')) {
+                $message .= sprintf('Did you add a "'.static::FIREWALL_OPTION.'" key under your "%s" firewall?', $firewallName);
```
```suggestion
                $message .= sprintf(' Did you forget to add a "'.static::FIREWALL_OPTION.'" key under your "%s" firewall?', $firewallName);
```

chalasr commented on src/Symfony/Bundle/SecurityBundle/Security/FirewallAwareTrait.php:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @internal
+ */
+trait FirewallAwareTrait
```
I'm not sure this is worth the added complexity/duck typing/readability loss

wouterj replied:
I'm happy to revert this change. As there are quite some checks (3 if statements + error messages) and it's used by 3 classes if this PR is merged, I thought it was nice to make a trait for it (to not forget a check). But I agree the implementation is a bit weird to still generate the correct error messages.

Nyholm replied:
Im happy with the trait.

chalasr commented on src/Symfony/Component/Security/Http/EventListener/RememberMeConditionsListener.php:
```diff
@@ -0,0 +1,79 @@
+ * @final
+ * @experimental in 5.3
+ */
+class RememberMeConditionsListener implements EventSubscriberInterface
```
`CheckRememberMeConditionsListener` to make it clear what it does?

chalasr commented on src/Symfony/Component/Security/Http/RememberMe/AbstractRememberMeHandler.php:
```diff
@@ -0,0 +1,121 @@
+        try {
+            $user = $this->userProvider->loadUserByUsername($rememberMeDetails->getUsername());
+        } catch (\Exception $e) {
+            if (!$e instanceof AuthenticationException) {
```
Throwing something else that an `AuthenticationException` means transgressing the `loadUserByUsername()`' contract. We should not silently fix such bad logic (not sure we should care at all)

wouterj replied:
This was a copy-past from the current remember me system, but I agree. Let's fix it

chalasr commented on src/Symfony/Component/Security/Http/RememberMe/RememberMeDetails.php:
```diff
@@ -0,0 +1,85 @@
+    {
+        $cookieParts = explode(self::COOKIE_DELIMITER, base64_decode($rawCookie), 4);
+        if (false === $cookieParts[1] = base64_decode($cookieParts[1], true)) {
+            throw new AuthenticationException('Username contains a character from outside the base64 alphabet.');
```
`BadCredentialsException`?

wouterj replied:
I think this is an impossible code path (and purely added as the current remember me system also does this check and strictly, `base64_encode()` can return `false`). As there is nothing a user can do about this, I don't think a `BadCredentialsException` makes sense here (but I'm happy to hear other ways `BadCredentialsException` could be useful).

For reference, it's `AuthenticationException` in the 5.2 code: https://github.com/symfony/symfony/blob/f052d11a43f00c9d0233033954657743fca8df0a/src/Symfony/Component/Security/Http/RememberMe/TokenBasedRememberMeServices.php#L39-L41

chalasr commented on src/Symfony/Component/Security/Core/Signature/ExpiredSignatureStorage.php:
```diff
@@ -9,14 +9,14 @@
+ * @author Ryan Weaver <ryan@symfonycasts.com>
  */
-class ExpiredLoginLinkStorage
+final class ExpiredSignatureStorage
```
This class has been introduced in 5.2 and is not experimental. We need a BC layer for this move, isn't it?

wouterj replied:
This class was introduced as part of the login link package, which is only supported by the authenticator-based security. So I think it was just forgotten to mark this class as experimental. @weaverryan can you confirm here?

chalasr replied:
Fine for me 👍 Shall we add the missing `experimental` flag to the new class then?

chalasr reviewed (changes requested):
Looks great!

fabpot commented:
@weaverryan Can you do a last review before we merge?

derrabus commented:
A rebase is needed.

ro0NL commented on src/Symfony/Bundle/SecurityBundle/Security/FirewallAwareTrait.php:
```diff
@@ -0,0 +1,52 @@
+
+    private function getForFirewall(): object
+    {
+        $getServiceIdentifier = function () { return str_replace('FirewallAware', '', static::class); };
```
why the callable?

ro0NL commented on src/Symfony/Component/Security/Http/RememberMe/RememberMeHandlerInterface.php:
```diff
@@ -0,0 +1,53 @@
+     * The actual cookie should be set as an attribute on the master request,
+     * which is transformed into a response cookie by {@see ResponseListener}.
+     */
+    public function createRememberMeCookie(UserInterface $user): void;
```
why not returning a `RememberMeDetails` holding the cookie, to be leveraged for `$request->attributes->set(ResponseListener::COOKIE_ATTR_NAME, `

IIUC each handler needs to anticipate this now?

wouterj replied:
I prefer this method to be fully self-contained, as it'll be used outside the Symfony framework (e.g. it's used by the SchebTwoFactorBundle to create the correct cookie after completing the 2fa).

ro0NL replied:
> it'll be used outside the Symfony framework

isnt that even more reason to solely expose the HTTP cookie here?

AFAIK we could even aim to avoid using the COOKIE_ATTR_NAME, as well as using the request for glueing.

chalasr reviewed (commented):
`master request` needs to be replaced by `main request` everywhere after #40536

weaverryan commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -148,15 +182,33 @@ public function addConfiguration(NodeDefinition $node)
+                    ->ifString()->then(function ($v) { return ['service' => $v]; })
+                ->end()
+                ->children()
+                    ->scalarNode('service')->end()
```
Should we add an `->info()` to this?

weaverryan commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -94,31 +98,61 @@ public function create(ContainerBuilder $container, string $id, array $config, ?
+        // create remember me handler (which manage the remember me cookies)
+        $rememberMeHandlerId = 'security.authenticator.remember_me_handler.'.$firewallName;
+        if (isset($config['service'])) {
+            $container->setDefinition($rememberMeHandlerId, $container->getDefinition($config['service']))
```
Hmm. `$container` is just the `ContainerBuilder` that's passed to `SecurityExtension`, right? So a user's custom service definition would not exist in this container.

We would normally use an alias instead... but I don't think we can set a tag on the alias. Another option would be to register a service from the security component that just decorates the user's... though on a high level, that's sort of just hacking around the problem with an extra class that shouldn't really be needed.

wouterj replied:
Hmm, this tag is at the moment only used for discovery by the doctrine schema listener. Maybe we should just ignore it for custom handlers?

weaverryan commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -94,31 +98,61 @@ public function create(ContainerBuilder $container, string $id, array $config, ?
+        if (isset($config['service'])) {
+            $container->setDefinition($rememberMeHandlerId, $container->getDefinition($config['service']))
+                ->addTag('security.remember_me_handler', ['firewall' => $firewallName]);
+        } elseif (isset($config['token_provider'])) {
```
Minor... but it would be ideal if there were validation to prevent `service` AND `token_provider` from both being set :)

weaverryan commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -217,17 +268,29 @@ private function createRememberMeServices(ContainerBuilder $container, string $i
-            ->addTag('kernel.event_subscriber', ['dispatcher' => 'security.event_dispatcher.'.$id])
-            ->replaceArgument(0, new Reference($rememberMeServicesId))
-        ;
+        $tokenProviderId = $config['service'];
```
The `service` is key is not necessarily set. If use the config from the PR description:

```
            remember_me:
                secret: '%kernel.secret%'
                token_provider:
                    doctrine: true
```

You currently get:

> Warning: Undefined array key "service"

weaverryan commented on src/Symfony/Bridge/Doctrine/SchemaListener/RememberMeTokenProviderDoctrineSchemaSubscriber.php:
```diff
@@ -0,0 +1,64 @@
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+final class RememberMeTokenProviderDoctrineSchemaSubscriber implements EventSubscriber
```
> Furthermore, a schema listener is created. Whenever the doctrine token provider is used, 
> make:migration/doctrine:schema:update will automatically create the required table.

I think this is not true quite yet - I don't think this class is registered as a service anywhere - so I don't get any automatic schema updates after enabling Doctrine.

wouterj replied:
Registering it is DoctrineBundle's job, see https://github.com/doctrine/DoctrineBundle/pull/1297

weaverryan replied:
doh! Of course! I will bring in that PR to my project to keep testing

weaverryan replied:
It's working nicely with that PR :)

weaverryan commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -217,17 +268,29 @@ private function createRememberMeServices(ContainerBuilder $container, string $i
+        $tokenProviderId = $config['service'];
+        if ($config['doctrine']['enabled'] ?? false) {
+            if (!class_exists(DoctrineTokenProvider::class)) {
+                throw new InvalidConfigurationException('"remember_me.token_provider.doctrine" cannot be enabled as symfony/doctrine-bridge is not installed.');
```
```suggestion
                throw new InvalidConfigurationException('Cannot use the "doctrine" token provider for remember_me because symfony/doctrine-bridge is not installed.');
```

weaverryan commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -565,6 +566,10 @@ private function createContextListener(ContainerBuilder $container, string $cont
+        if (null !== $firewallEventDispatcherId) {
+            $listener->replaceArgument(4, new Reference($firewallEventDispatcherId));
+            $listener->addTag('kernel.event_listener', ['event' => KernelEvents::RESPONSE, 'method' => 'onKernelResponse']);
+        }
```
Can you explain this piece? Currently, when `ContextListener::authenticate()` is called, that class subscribes to the `KernelEvents:RESPONSE` event. I'm actually not sure what the reason is for that odd setup.

I believe you added this code because we are NOW passing in the firewall-specific dispatcher into `ContextListener`, and not the *main* dispatcher. And so, subscribing to `KernelEvents:RESPONSE` no longer has any effect. That makes sense. Two questions though:

1) Why do we have the current, weird setup where we register the listener inside the class? If that is there for a legitimate reason, then why is it ok for us to NOW register it in a more traditional way?

2) Should (can we?) deprecate the old functionality of having `ContextListener` register the listener? It's a bit odd (and just ugly code to keep around) for `ContextListener` to register the listener on the firewall dispatcher... which will never actually have any effect.

wouterj replied:
See the discussion in a previous comment on this PR: https://github.com/symfony/symfony/pull/40145#discussion_r574802604

Basically, we must only do something in `onKernelResponse()` if `authenticate()` was called for the same request (see referenced PRs in the previous link). Previously, that was "solved" by dynamically registering this listener. For performance reasons, and to make this less weird, I've refactored it to instead rely on saving the state in the request attributes.

wouterj replied:
> Should (can we?) deprecate the old functionality of having ContextListener register the listener?

I don't think we can check whether `onKernelResponse()` listener was registered by the `authenticate()` method or config. So I decided to just always register it in the 5.x, and then remove that line in 6.0. That's a very small hard break (if someone used the `ContextListener` together with the EventDispatcher component outside the framework), but as it is a break, doing it in 6.0 seems better to me.

weaverryan commented on src/Symfony/Component/Security/Http/Authenticator/RememberMeAuthenticator.php:
```diff
@@ -57,32 +65,37 @@ public function supports(Request $request): ?bool
-        if ($request->attributes->has('_remember_me_token')) {
-            return null;
+        if (($cookie = $request->attributes->get(ResponseListener::COOKIE_ATTR_NAME)) && null === $cookie->getValue()) {
+            return false;
```
What condition is this catching? Is it where, earlier in this request, `clearRememberMeCookie()` was called?

wouterj replied:
Yes, e.g. when the user is deauthenticated upon refresh, they should not be authenticated again by the remember me system. See the (fixed) bug reported in https://github.com/symfony/symfony/issues/35198

weaverryan commented on src/Symfony/Component/Security/Http/Authenticator/RememberMeAuthenticator.php:
```diff
@@ -57,32 +65,37 @@ public function supports(Request $request): ?bool
-            throw new \LogicException('No remember me token is set.');
+        $rawCookie = $request->attributes->get(ResponseListener::COOKIE_ATTR_NAME);
+        if (!$rawCookie || null === $rawCookie->getValue()) {
+            $rawCookie = $request->cookies->get($this->cookieName);
```
I'm not quite following the logic here. We look for the cookie in `$request->attributes` first and THEN in `$request->cookies`. Why? The flow may just need some clues with short comments

wouterj replied:
I'm not sure if I did this with a reason, or if this is just because that's also how it's done in the "legacy" remember me: https://github.com/symfony/symfony/blob/9abeb248a58d22d94f18c1055f648fdf1a4c80e5/src/Symfony/Component/Security/Http/RememberMe/AbstractRememberMeServices.php#L101-L107

wouterj replied:
Turns out that this is great catch. It's based on a wrong copy-paste from the services to the authenticator. `authenticate()` should only use the cookie (the null attribute case is covered by `support()` and is required for scenarios as described in https://github.com/symfony/symfony/issues/35198 )

weaverryan commented on src/Symfony/Component/Security/Http/Firewall/ContextListener.php:
```diff
@@ -128,10 +131,17 @@ public function authenticate(RequestEvent $event)
+
+                if ($this->rememberMeServices) {
+                    $this->rememberMeServices->loginFail($request);
+                }
```
Would passing `RememberMeServices` to this class be deprecated later?

wouterj replied:
Yes, but we can only do so when deprecating the legacy system (as this is completely valid in the legacy system).

weaverryan commented on src/Symfony/Component/Security/Http/Firewall/ContextListener.php:
```diff
@@ -159,11 +169,13 @@ public function onKernelResponse(ResponseEvent $event)
         $request = $event->getRequest();
 
-        if (!$request->hasSession()) {
+        if (!$request->hasSession() || !$request->attributes->get('_security_firewall_run', false)) {
```
Ah, this relates to my question earlier about how the listener is registered.

wouterj replied:
Indeed :)

weaverryan reviewed (commented):
I still have a few more cases to test, but I wanted to get feedback going ASAP. This looks really nice - no surprise :). Thanks Wouter!

weaverryan commented on src/Symfony/Component/Security/Http/RememberMe/PersistentRememberMeHandler.php:
```diff
@@ -0,0 +1,110 @@
+     */
+    public function processRememberMe(RememberMeDetails $rememberMeDetails, UserInterface $user): void
+    {
+        [$series, $tokenValue] = explode(':', $rememberMeDetails->getValue());
```
There's an annoying edge-case here, I hit it when switching from the signature-based remember me to persistent-based. I was already logged in with a remember me cookie and refresh. This, of course, SHOULD log me out - as I got the remember me cookie using the "signature-based" and I just switched to "persistent". But I get a hard error:

> Warning: Undefined array key 1

Coming from this line, since the value does not have any `:` in it.

I added this to the top of the function and it all worked fine:

```php
        if (!str_contains($rememberMeDetails->getValue(), ':')) {
            throw new AuthenticationException('The cookie is incorrectly formatted');
        }
```

weaverryan commented on src/Symfony/Component/Security/Http/RememberMe/PersistentRememberMeHandler.php:
```diff
@@ -0,0 +1,110 @@
+            throw new CookieTheftException('This token was already used. The account is possibly compromised.');
+        }
+
+        if ($persistentToken->getLastUsed()->getTimestamp() + $this->options['lifetime'] < time()) {
```
Does this mean that if I just continually "use" my cookie that I could get it to last longer than its lifetime? What is the meaning of lifetime? Like, if I set it to 1 week.. should it absolutely not be usable in 1 week? Or something different?

wouterj replied:
A new remember me cookie is created each time the remember me is called (i.e. the session has expired). The ttl defines the time to the life from the last usage of the remember me. So e.g. 1 week means you're fully logged out after 1 week of inactivity, but you can continually use it if you use the website on a daily basis.

This is btw an exact copy of the current `PersistentTokenBasedRememberMeServices` logic.

weaverryan replied:
Ok, so this matches past functionality. So that's ok.

But I also think, perhaps not in this PR (or perhaps yes in this PR?) we should change that. When I set my remember me lifetime to 1 week, I would expect that it is valid for 1 week MAX - that a user (especially a bad user & especially when using non-persistent cookies) couldn't continue to re-use that indefinitely to access my account. It seems to me that the correct behavior is:

A) With persistent storage, we should store a `createdAt` date and use THAT when comparing the lifetime, not a `lastUsed` date.
B) For non-persistent storage (and i guess this probably also applies to persistent) the cookie should contain the lifetime / expiration in it. And this expiration should not *extend* when I use the cookie (even if we need to create a new remember me cookie when the session is expired, we should do so with a *shorter* expiration date, so that the new cookie expires at the same date as the original).

This is an important, and potentially controversial (? ) point. So I would be ok with merging the PR (as it matches the existing functionality) and opening an issue / PR to address this after.

wouterj replied:
Hmm, let's indeed discuss this in an issue.

Looking at e.g. GitLab, they implement Remember me like we do currently. So I'm not sure which of the 2 is expected behavior.

weaverryan commented on src/Symfony/Component/Security/Http/RememberMe/PersistentRememberMeHandler.php:
```diff
@@ -0,0 +1,110 @@
+        parent::clearRememberMeCookie();
+
+        $cookie = $this->requestStack->getMasterRequest()->cookies->get($this->options['name']);
+        if (null === $cookie) {
```
What is this check for? If the request doesn't have the cookie, we *avoid* deleting it from the storage?

wouterj replied:
We need the cookie of the request to know which one to delete from the storage. In theory, `ParameterBag::get()` may return null, so that's what this check is taking care of.

weaverryan commented on src/Symfony/Component/Security/Http/RememberMe/RememberMeHandlerInterface.php:
```diff
@@ -0,0 +1,53 @@
+     * Validates the remember me cookie and returns the associated User.
+     *
+     * This method should also invalidate the current cookie, if this is
+     * stored in a different place (e.g. a database).
```
Can you explain this more? Are you referring to the remember me cookie? If so, why would we need to invalidate it? Is this what we're doing in, for example, `PersistentRememberMeHandler::processRememberMe()` when we call `$this->createCookie()`?

wouterj replied:
Updated the PHPdocs

weaverryan commented on src/Symfony/Component/Security/Http/RememberMe/AbstractRememberMeHandler.php:
```diff
@@ -0,0 +1,117 @@
+    /**
+     * @throws AuthenticationException
+     */
+    abstract protected function processRememberMe(RememberMeDetails $rememberMeDetails, UserInterface $user): void;
```
We could use some docs on this, especially since, I think, it's required that this method eventually calls `$this->createCookie()` to (I think?) create a fresh cookie.

wouterj replied:
Updated the PHPdocs

weaverryan reviewed (commented):
I've left some comments, but i've tested this thoroughly, including the signature (non-persistent) remember me, persistent remember, watching the persistent tokens get removed on logout (including if the user "changes"). I've also tested the login_link to make sure that is still functioning. Things are working *beautifully*!

weaverryan commented:
Also, I've just tested basic remember me functionality (getting the cookie set, invalidating the session, watching the remember me cookie log me back in, removal of cookie on log out) on the OLD security system and it appears to be working just fine :)

Nyholm commented on src/Symfony/Bridge/Doctrine/SchemaListener/RememberMeTokenProviderDoctrineSchemaSubscriber.php:
```diff
@@ -0,0 +1,64 @@
+                continue;
+            }
+
+            $tokenProvider->configureSchema($event->getSchema(), $dbalConnection);
```
Fancy computer scientists would say that we should use "De Morgan's law" here. What that really means is to rewrite the logic so it means the same but expressed in the opposite way. 

I think it would be easier to read. 


```php
if (
    $rememberMeHandler instanceof PersistentRememberMeHandler
    && ($tokenProvider = $rememberMeHandler->getTokenProvider()) instanceof DoctrineTokenProvider
) {
    $tokenProvider->configureSchema($event->getSchema(), $dbalConnection);
}
```

Nyholm commented on src/Symfony/Bundle/SecurityBundle/RememberMe/DecoratedRememberMeHandler.php:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @internal
+ */
+final class DecoratedRememberMeHandler implements RememberMeHandlerInterface
```
If we remove the old "remember me" system in 6.0. Do we still need this class in 6.0?

wouterj replied:
Yes, see https://github.com/symfony/symfony/pull/40145#discussion_r599703158

Nyholm commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -94,31 +99,66 @@ public function create(ContainerBuilder $container, string $id, array $config, ?
     {
-        $templateId = $this->generateRememberMeServicesTemplateId($config, $firewallName);
-        $rememberMeServicesId = $templateId.'.'.$firewallName;
+        if (!$container->hasDefinition('security.authenticator.remember_me')) {
```
Why do we need this check?

wouterj replied:
Previously, security loaded all services in the main DI config of the SecurityBundle. We're now only loading services whenever a factory is used (to avoid "useless" services in the dev environment)

Nyholm commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -94,31 +99,66 @@ public function create(ContainerBuilder $container, string $id, array $config, ?
+        // create remember me handler (which manage the remember me cookies)
+        $rememberMeHandlerId = 'security.authenticator.remember_me_handler.'.$firewallName;
+        if (isset($config['service']) && isset($config['token_provider'])) {
+            throw new InvalidConfigurationException(sprintf('You must configure either "service" or "token_provider", but both are set at "security.firewalls.%s.remember_me".', $firewallName));
```
You may also configure none. 

> You cannot use both "service" and "token_provider", but both are set at "security.firewalls.%s.remember_me".

Nyholm commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -148,15 +188,33 @@ public function addConfiguration(NodeDefinition $node)
+                ->prototype('scalar')->end()
+                ->requiresAtLeastOneElement()
+                ->info('An array of properties on your User that are used to sign the rememberme cookie. If any of these change, all existing cookies will become invalid.')
+                ->example(['email', 'password'])
```
I like this description!

Nyholm commented on src/Symfony/Component/Security/Http/Authenticator/RememberMeAuthenticator.php:
```diff
@@ -98,7 +107,15 @@ public function onAuthenticationSuccess(Request $request, TokenInterface $token,
+            } elseif ($exception instanceof UnsupportedUserException) {
+                $this->logger->warning('User class for remember-me cookie not supported.', ['exception' => $exception]);
+            } elseif (!$exception instanceof CookieTheftException) {
+                $this->logger->debug('Remember-Me authentication failed.', ['exception' => $exception]);
```
We use `remember-me`, `rememberme`, `Remember-Me` and `remember_me`, it sometimes seams arbitrary when and how. Im sure there is a system and you dont need to explain it for my sake. 

But can you double check this usage is correct?

wouterj replied:
Thanks for your trust, but no there was no system here (just a mixture of copy-pasting and writing new variants).

I've updated (only in the new system code) everything to talk about "remember me system" and "remember-me cookie/token".

Nyholm commented on src/Symfony/Component/Security/Http/EventListener/CheckRememberMeConditionsListener.php:
```diff
@@ -0,0 +1,79 @@
+
+        /** @var RememberMeBadge $badge */
+        $badge = $passport->getBadge(RememberMeBadge::class);
+        if (!$this->options['always_remember_me']) {
```
Another minor: 

```php
if ($this->options['always_remember_me']) {
  $badge->enable();
  return;
}

$parameter = ParameterBagUtils::getRequestParameterValue($event->getRequest(), $this->options['remember_me_parameter']);
if ('true' === $parameter || 'on' === $parameter || '1' === $parameter || 'yes' === $parameter || true === $parameter) {
    $badge->enable();
} elseif (null !== $this->logger) {
    $this->logger->debug('Remember me disabled; request does not contain remember me parameter ("{parameter}").', ['parameter' => $this->options['remember_me_parameter']]);
}
```

Nyholm reviewed (commented):
Thank you. 
I did my best reviewing this. I just had some minor comments and suggestions.

weaverryan reviewed (approved):
This looks ready to go 👍

chalasr commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/ReplaceDecoratedRememberMeHandlerPass.php:
```diff
@@ -0,0 +1,61 @@
+    private static $handlerServiceTag = 'security.remember_me_handler';
+
+    /**
+     * {@inheritDoc}
```
```suggestion
     * {@inheritdoc}
```

chalasr commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/ReplaceDecoratedRememberMeHandlerPass.php:
```diff
@@ -0,0 +1,61 @@
+ */
+final class ReplaceDecoratedRememberMeHandlerPass implements CompilerPassInterface
+{
+    private static $handlerServiceTag = 'security.remember_me_handler';
```
```suggestion
    private const HANDLER_TAG = 'security.remember_me_handler';
```

chalasr commented on src/Symfony/Component/Security/Core/Signature/Exception/ExpiredSignatureException.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class ExpiredSignatureException extends \RuntimeException
```
this should extend the security-core RuntimeException

chalasr commented on src/Symfony/Component/Security/Core/Signature/Exception/InvalidSignatureException.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class InvalidSignatureException extends \RuntimeException
```
same

chalasr commented on src/Symfony/Component/Security/Http/LoginLink/LoginLinkHandler.php:
```diff
@@ -29,25 +31,18 @@ final class LoginLinkHandler implements LoginLinkHandlerInterface
+    private $signatureHashUtil;
 
-    public function __construct(UrlGeneratorInterface $urlGenerator, UserProviderInterface $userProvider, PropertyAccessorInterface $propertyAccessor, array $signatureProperties, string $secret, array $options, ?ExpiredLoginLinkStorage $expiredStorage)
+    public function __construct(UrlGeneratorInterface $urlGenerator, UserProviderInterface $userProvider, SignatureHasher $signatureHashUtil, array $options)
```
this breaking change could be mentioned in the changelog

wouterj commented:
Thanks for the detailed review @chalasr :)

chalasr commented:
Thank you @wouterj.

scheb commented:
@wouterj Congratulations for having this merged! 🥳 Thank you for all of your work and thank you to everyone who gave feedback on this thread. I believe it's a great step forward!

PhilETaylor commented:
cross posting https://github.com/scheb/2fa/issues/75 for visibility :-) where a live app, upgraded to `v5.3.0-BETA1` (locally, dont worry) now has a problem with the decorator and interface type hints.

scheb commented:
> cross posting [scheb/2fa#75](https://github.com/scheb/2fa/issues/75) for visibility :-) where a live app, upgraded to `v5.3.0-BETA1` (locally, dont worry) now has a problem with the decorator and interface type hints.

As I mentioned on scheb/2fa#75, just to let the people here know, that's nothing to worry about since I've been in the loop from the beginning of this PR here and I have an (unfinished) branch preparred to make scheb/2fa-bundle compatible with Symfony 5.3. Whoever wants to test Symfony 5.3 with 2fa-bundle is welcome to use that branch.

PhilETaylor commented:
Sorry :-) was too eager to test ... thanks everyone for the hard work - this looks amazing!

zerkms commented:
What was the reason to serialise `userFqcn` in the cookie? It's not used anywhere and leaks application implementation details for no good reason. The same for `expires`.

derrabus commented:
@zerkms Would you mind opening a new issue for this topic? Only few people monitor comments of already closed PRs.
